File: app/src/main/java/com/example/digitalmonk/service/vpn/heartbeat/VpnHeartBeatEntity.java
package com.example.digitalmonk.service.vpn.heartbeat;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

/**
 * Why we made this file:
 * This Room Entity defines the SQLite table structure for the VPN heartbeat log.
 * By writing "ALIVE" every 7 minutes, the external Watchdog worker can query this
 * table and know if the Android OS silently killed the VPN. If the last record
 * isn't "ALIVE" or is too old, the Watchdog knows it needs to revive the VPN.
 */
@Entity(tableName = "vpn_heartbeat")
public class VpnHeartBeatEntity {

    public static final String TYPE_ALIVE = "ALIVE";
    public static final String TYPE_STOPPED = "STOPPED";

    @PrimaryKey(autoGenerate = true)
    private long id;

    private String type;
    private long timestamp;

    /**
     * Default Constructor for Room.
     */
    public VpnHeartBeatEntity() {}

    /**
     * Convenience constructor mimicking Kotlin's default timestamp parameter.
     */
    public VpnHeartBeatEntity(String type) {
        this.type = type;
        this.timestamp = System.currentTimeMillis();
    }

    /**
     * Full constructor.
     */
    public VpnHeartBeatEntity(String type, long timestamp) {
        this.type = type;
        this.timestamp = timestamp;
    }

    public boolean isAlive() {
        return TYPE_ALIVE.equals(type);
    }

    // ── Getters and Setters Required by Room ──────────────────────────────────

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }
}

File: app/src/main/java/com/example/digitalmonk/service/vpn/heartbeat/VpnHeartBeatEntity.kt
package com.example.digitalmonk.service.vpn.heartbeat

import androidx.room.Dao
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query

// ── Entity ────────────────────────────────────────────────────────────────────

/**
 * Persists a heartbeat record every 7 minutes while DnsVpnService is running.
 * On clean shutdown, a STOPPED record is written so the watchdog knows it was intentional.
 *
 * Pattern borrowed from DDG's VpnServiceHeartbeat.kt:
 *   app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/heartbeat/
 */
@Entity(tableName = "vpn_heartbeat")
data class VpnHeartBeatEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val type: String,           // ALIVE or STOPPED
    val timestamp: Long = System.currentTimeMillis()
) {
    fun isAlive(): Boolean = type == TYPE_ALIVE

    companion object {
        const val TYPE_ALIVE   = "ALIVE"
        const val TYPE_STOPPED = "STOPPED"
    }
}

// ── DAO ───────────────────────────────────────────────────────────────────────

@Dao
interface VpnHeartBeatDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(heartBeat: VpnHeartBeatEntity)

    /**
     * Returns the most recent heartbeat record, or null if none exist.
     * Used by VpnMonitorService.onDestroy() and VpnHeartbeatMonitorWorker
     * to decide whether to restart the VPN.
     */
    @Query("SELECT * FROM vpn_heartbeat ORDER BY timestamp DESC LIMIT 1")
    fun getLastHeartBeat(): VpnHeartBeatEntity?

    /** Cleans up old records — keep only the latest 20 to avoid unbounded growth. */
    @Query("DELETE FROM vpn_heartbeat WHERE id NOT IN (SELECT id FROM vpn_heartbeat ORDER BY timestamp DESC LIMIT 20)")
    fun pruneOldRecords()
}